File: app/routers/tasks.py
from fastapi import APIRouter, HTTPException, Depends, status
from sqlalchemy.orm import Session
from app.auth.dependencies import get_current_user
from uuid import UUID
import logging

from app.models.task import TaskPublic, TaskCreate, TaskUpdate, TaskStatus
from app.models.user import User
from app.db.session import get_db
from app.storage.db_tasks import create_task, list_tasks, get_task_by_id, update_task, delete_task
from app.api.serializers import task_orm_to_public

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/tasks", tags=["tasks"])



@router.get("", response_model=list[TaskPublic])
def list_tasks_endpoint(
    current_user: User = Depends(get_current_user), 
    db: Session = Depends(get_db)
    ) ->list[TaskPublic]:
    """
    Task listing endpoint.
    - Retrieve list of all tasks from database (specific to authenticated user)
    - Serialize to TaskPublic and return
    """
    task_list_public = list()
    logger.info("Fetching tasks for user_id=%s", str(current_user.id))
    task_list = list_tasks(db, user_id=current_user.id)
    for task in task_list:
        public = task_orm_to_public(task)
        task_list_public.append(public)
    return task_list_public



@router.post("", status_code=status.HTTP_201_CREATED, response_model=TaskPublic)
def create_task_endpoint(
    data: TaskCreate, 
    current_user: User = Depends(get_current_user), 
    db: Session = Depends(get_db)
    ) -> TaskPublic:
    """
    Create new task endpoint.
    - Call task creation method
    - Return task
    """
    logger.info("Creating task with title='%s'", data.title)
    task = create_task(
        db,
        owner_id=current_user.id,
        title=data.title, 
        description=data.description,
        status=data.status.value,
        due_date=data.due_date
        )
    logger.info("Task created. id=%s", str(task.id))
    task_public = task_orm_to_public(task)
    return task_public



@router.get("/{task_id}", response_model=TaskPublic)
def get_task_by_id_endpoint(
    task_id: UUID, 
    current_user: User = Depends(get_current_user), 
    db: Session = Depends(get_db)
    ) -> TaskPublic:
    """
    Fetch a task by id endpoint.
    - Call task getter method
    - Return task
    """
    logger.info("Fetching task id=%s", str(task_id))
    task = get_task_by_id(
        db, 
        task_id=task_id, 
        user_id=current_user.id
        )
    if not task:
        logger.warning("Task not found id=%s", str(task_id))
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
    task_public = task_orm_to_public(task)
    return task_public



@router.patch("/{task_id}", response_model=TaskPublic)
def update_task_endpoint(
    task_id: UUID, 
    data: TaskUpdate, 
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
    ) -> TaskPublic:
    """
    Update an existing task endpoint.
    - Call task update endpoint
    - Return task
    - If no task matching id exists, raise HTTPException
    """
    logger.info("Updating task id=%s", str(task_id))
    task = update_task(
        db, 
        task_id=task_id, 
        data=data,
        owner_id=current_user.id
        )
    if not task:
        logger.warning("Task not found for update id=%s", str(task_id))
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
    logger.info("Task updated id=%s", str(task_id))
    task_public = task_orm_to_public(task)
    return task_public



@router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_task_endpoint(
    task_id: UUID, 
    current_user: User = Depends(get_current_user), 
    db: Session = Depends(get_db)
    ):
    """
    Delete a task endpoint.
    - Call task deletion method
    - If no task matching id exists, raise HTTPException
    """
    logger.info("Deleting task id=%s", str(task_id))
    deleted = delete_task(db, task_id=task_id, user_id=current_user.id)
    if not deleted:
        logger.warning("Task not found for delete id=%s", str(task_id))
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
    logger.info("Task deleted id=%s", str(task_id))

File: frontend/src/lib/api/tasks.ts
export type Task = {
    id: string;
    title: string;
    description?: string | null;
    status: "pending" | "in_progress" | "completed";
    due_date?: string | null;
    created_at: string;
    updated_at?: string | null;
};

export type TaskUpdate = {
    title?: string;
    description?: string;
    status?: "pending" | "in_progress" | "completed";
    due_date?: string;
}

export type TaskCreate = {
    title: string;
    description?: string;
    status: "pending" | "in_progress";
    due_date?: string;
}

export async function getTaskById(task_id: string): Promise<Task | null> {
    const access_token = localStorage.getItem("access_token");
    try {
        const response = await fetch(`${import.meta.env.VITE_API_URL}/api/tasks/${task_id}`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${access_token}`
            }
        });
        if (!response.ok) {
            console.error(`Error fetching task ${task_id}`);
            return null;
        }
        const task: Task = await response.json();
        return task
    } catch(error) {
        console.error(`Error fetching task ${task_id}`, error);
        return null;
    }
}

export async function deleteTask(task_id: string): Promise<boolean> {
    const access_token = localStorage.getItem("access_token");
    try {
        const response = await fetch(`${import.meta.env.VITE_API_URL}/api/tasks/${task_id}`, {
            method: 'DELETE',
            headers: {
                'Authorization': `Bearer ${access_token}`,
            },
        })
        if (!response.ok) return false;
        return true;
    } catch(error) {
        console.error("Error when deleting task: ", error);
        return false;
    }
}

export async function createTask(task: TaskCreate): Promise<Task | null> {
    const access_token = localStorage.getItem("access_token");
    try {
        const response = await fetch(`${import.meta.env.VITE_API_URL}/api/tasks`, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${access_token}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(task),
        });
        if (!response.ok) {
            return null;
        }
        const updatedTask = await response.json();
        return updatedTask;
    } catch(error) {
        console.error("Error while creating task: ", error);
        return null;
    }
}

export async function getTasks(): Promise<Task[] | null> {
    const access_token = localStorage.getItem("access_token");
    try {
        const response = await fetch(`${import.meta.env.VITE_API_URL}/api/tasks`, {
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${access_token}`
            }
        });
        if (!response.ok) {
            console.error("Error fetching tasks.");
            return null;
        }
        const taskList: Task[] = await response.json();
        return taskList;
    } catch (error) {
        console.error("Error fetching tasks: ", error);
        return null;
    }
}

export async function updateTask(
    task_id: string,
    changes: TaskUpdate,
): Promise<Task | null> {
    const access_token = localStorage.getItem("access_token");
    try {
        const response = await fetch(`${import.meta.env.VITE_API_URL}/api/tasks/${task_id}`, {
            method: 'PATCH',
            headers: {
                'Authorization': `Bearer ${access_token}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(changes),
        });
        if (!response.ok) {
            return null;
        }
        const task: Task = await response.json();
        return task;
    } catch(error) {
        console.error("Error updating task: ", error);
        return null;
    }
}

export async function setTaskStatus(
    task_id: string,
    status: "pending" | "in_progress" | "completed",
): Promise<Task | null> {
    const access_token = localStorage.getItem("access_token");
    const update: TaskUpdate = { status: status }
    try {
        const response = await fetch(`${import.meta.env.VITE_API_URL}/api/tasks/${task_id}`, {
            method: 'PATCH',
            headers: {
                'Authorization': `Bearer ${access_token}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(update),
        });
        if (!response.ok) {
            return null;
        }
        const task: Task = await response.json();
        return task;
    } catch(error) {
        console.error("Error updating task: ", error);
        return null;
    }
}